packages/node-core/src/indexer/poi/poi.service.ts

```diff
@@ -1,55 +1,261 @@
 // Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
 // SPDX-License-Identifier: GPL-3.0
 
-import {isMainThread} from 'node:worker_threads';
 import {Injectable, OnApplicationShutdown} from '@nestjs/common';
+import {EventEmitter2} from '@nestjs/event-emitter';
+import {DEFAULT_FETCH_RANGE, delay, POI_AWAIT_TIME} from '@subql/common';
 import {hexToU8a} from '@subql/utils';
+import {QueryTypes} from '@subql/x-sequelize';
+import {PoiEvent} from '../../events';
+import {getLogger} from '../../logger';
+import {ProofOfIndex, SyncedProofOfIndex} from '../entities/Poi.entity';
 import {StoreCacheService} from '../storeCache';
 import {CachePoiModel} from '../storeCache/cachePoi';
+import {PoiBlock} from './PoiBlock';
 
-const DEFAULT_PARENT_HASH = hexToU8a('0x00');
+const GENESIS_PARENT_HASH = hexToU8a('0x00');
+const logger = getLogger('ProofOfIndex');
+
+const syncingMsg = (start: number, end: number, size: number) =>
+  logger.info(`Synced POI [${start} - ${end}] Poi, total ${size} blocks `);
 
 @Injectable()
 export class PoiService implements OnApplicationShutdown {
   private isShutdown = false;
-  private latestPoiBlockHash?: Uint8Array | null;
-  private poiRepo?: CachePoiModel;
+  private _poiRepo?: CachePoiModel;
+  private _latestSyncedPoi?: ProofOfIndex;
+  private isSyncing = false;
+  private _projectId?: string;
 
-  constructor(private storeCache: StoreCacheService) {}
+  constructor(private storeCache: StoreCacheService, protected eventEmitter: EventEmitter2) {}
 
   onApplicationShutdown(): void {
     this.isShutdown = true;
   }
 
-  async init(): Promise<void> {
-    this.poiRepo = this.storeCache.poi ?? undefined;
-    this.latestPoiBlockHash = await this.getLatestPoiBlockHash();
+  get poiRepo(): CachePoiModel {
+    if (!this._poiRepo) {
+      throw new Error(`No poi repo inited`);
+    }
+    return this._poiRepo;
+  }
+
+  get projectId(): string {
+    if (!this._projectId) {
+      throw new Error(`No project Id inited`);
+    }
+    return this._projectId;
+  }
+
+  get latestSyncedPoi(): SyncedProofOfIndex {
+    if (!this._latestSyncedPoi) {
+      throw new Error('_latestSyncedPoi in Poi service is not inited');
+    }
+    if (this._latestSyncedPoi.parentHash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, parentHash is missing`);
+    }
+    if (this._latestSyncedPoi.hash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, hash is missing`);
+    }
```

Comment:
What are the chances of these happening? If a user faced this error can we give them some instructions how to resolve the issue?

Reply:
We don't have any feature to reset syncedPoi, once we have this feature, we could log here let user to reset to previous valid poi height, also complete https://github.com/subquery/subql/blob/bfe6de09f6743db219f5e46f2022bf41711feb3a/packages/node-core/src/indexer/poi/poi.service.ts#L72

Reply:
Can we treat this same as `_latestSyncedPoi` not exist?